cloud/pkg/taskmanager/nodeupgradecontroller/node_upgrade_controller.go

```diff
@@ -0,0 +1,351 @@
+	req, ok := taskMessage.Msg.(commontypes.NodeUpgradeJobRequest)
+	if !ok {
+		klog.Errorf("convert message to commontypes.NodeUpgradeJobRequest failed")
+		return nil
+	}
+	for _, node := range nodes {
+		if needUpgrade(node, req.Version) {
+			validateNodes = append(validateNodes, node)
+		}
+	}
+	return validateNodes
+}
+
+func (ndc *NodeUpgradeController) StageCompleted(taskID string, state api.State) bool {
+	taskFSM := NewUpgradeTaskFSM(taskID)
+	return taskFSM.TaskStagCompleted(state)
+}
+
+func (ndc *NodeUpgradeController) GetNodeStatus(name string) ([]v1alpha1.TaskStatus, error) {
+	nodeUpgrade, err := ndc.CrdClient.OperationsV1alpha1().NodeUpgradeJobs().Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return nodeUpgrade.Status.Status, nil
+}
+
+func (ndc *NodeUpgradeController) GetNodeVersion(name string) (string, error) {
+	node, err := ndc.Informer.Core().V1().Nodes().Lister().Get(name)
+	if err != nil {
+		return "", err
+	}
+	strs := strings.Split(node.Status.NodeInfo.KubeletVersion, "-")
+	return strs[2], nil
+}
+
+func (ndc *NodeUpgradeController) UpdateNodeStatus(name string, nodeStatus []v1alpha1.TaskStatus) error {
+	nodeUpgrade, err := ndc.CrdClient.OperationsV1alpha1().NodeUpgradeJobs().Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	status := nodeUpgrade.Status
+	status.Status = nodeStatus
+	err = patchStatus(nodeUpgrade, status, ndc.CrdClient)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func patchStatus(nodeUpgrade *v1alpha1.NodeUpgradeJob, status v1alpha1.NodeUpgradeJobStatus, crdClient crdClientset.Interface) error {
+	oldData, err := json.Marshal(nodeUpgrade)
+	if err != nil {
+		return fmt.Errorf("failed to marshal the old NodeUpgradeJob(%s): %v", nodeUpgrade.Name, err)
+	}
+	nodeUpgrade.Status = status
+	newData, err := json.Marshal(nodeUpgrade)
+	if err != nil {
+		return fmt.Errorf("failed to marshal the new NodeUpgradeJob(%s): %v", nodeUpgrade.Name, err)
+	}
+
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return fmt.Errorf("failed to create a merge patch: %v", err)
+	}
+
+	result, err := crdClient.OperationsV1alpha1().NodeUpgradeJobs().Patch(context.TODO(), nodeUpgrade.Name, apimachineryType.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	if err != nil {
+		return fmt.Errorf("failed to patch update NodeUpgradeJob status: %v", err)
+	}
+	klog.V(4).Info("patch upgrade task status result: ", result)
+	return nil
+}
+
+func (ndc *NodeUpgradeController) Start() error {
+	go ndc.startSync()
+	return nil
+}
+
+func (ndc *NodeUpgradeController) startSync() {
+	nodeUpgradeList, err := ndc.CrdClient.OperationsV1alpha1().NodeUpgradeJobs().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf(err.Error())
+		os.Exit(2)
+	}
+	for _, nodeUpgrade := range nodeUpgradeList.Items {
+		if &(nodeUpgrade.Status) != nil && fsm.TaskFinish(nodeUpgrade.Status.State) {
```

Comment:
`&(nodeUpgrade.Status) != nil` the address of a variable cannot be nil

Reply:
done